### ISSUE TITLE
Problem mapping callable to generic protocol

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1315]
**Describe the bug**

Cannot map a generic callable to a generic protocol

**To Reproduce**
See code example below. The function `mapi` returning a generic protocol gives type error while `mapi_` returning a generic callable does not give type error.

**Expected behavior**
I would expect both `mapi` and `mapi_` to not give type error since the `Callable[[AsyncObservable[TSource]], AsyncObservable[TResult]]` should be compatible with the protocol `Projection[TSource, TResult]`. Perhaps I have misunderstood how this works, but I feel that this must be a bug.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.

<img width="1160" alt="Screenshot 2020-12-26 at 12 53 32" src="https://user-images.githubusercontent.com/849479/103150938-708f8180-4779-11eb-80b7-2adb82e3c2db.png">

```python
import sys
from functools import reduce
from typing import Any, Callable, Generic, Iterable, Protocol, Tuple, TypeVar, overload

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
TSource = TypeVar("TSource")
TResult = TypeVar("TResult")
TOther = TypeVar("TOther")


class AsyncObservable(Generic[TSource]):
    pass


class Projection(Protocol[TSource, TResult]):
    def __call__(self, __source: AsyncObservable[TSource]) -> AsyncObservable[TResult]:
        raise NotImplementedError


class Zipper(Protocol[TResult]):
    def __call__(self, __source: AsyncObservable[TSource]) -> AsyncObservable[Tuple[TSource, TResult]]:
        raise NotImplementedError


@overload
def compose(__fn1: Callable[[A], B], __fn2: Callable[[B], C]) -> Callable[[A], C]:
    ...


def compose(*fns: Callable[[Any], Any]) -> Callable[[Any], Any]:
    def _compose(source: Any) -> Any:
        return reduce(lambda acc, f: f(acc), fns, source)

    return _compose


infinite: Iterable[int] = range(sys.maxsize)


def zip_seq(sequence: Iterable[TOther]) -> Zipper[TOther]:
    def _zip_seq(source: AsyncObservable[TSource]) -> AsyncObservable[Tuple[TSource, TOther]]:
        raise NotImplementedError

    return _zip_seq


def starmap(mapper: Callable[..., TResult]) -> Projection[Iterable[Any], TResult]:
    raise NotImplementedError


def mapi(mapper: Callable[[TSource, int], TResult]) -> Projection[TSource, TResult]:
    return compose(
        zip_seq(infinite),
        starmap(mapper),
    )


def mapi_(mapper: Callable[[TSource, int], TResult]) -> Callable[[AsyncObservable[TSource]], AsyncObservable[TResult]]:
    return compose(
        zip_seq(infinite),
        starmap(mapper),
    )
```

**VS Code extension or command-line**
Vscode extension

```
[Info  - 10:49:17 AM] Pyright language server 1.1.96 starting
```

**Comment by erictraut** [comment]
You are correct that these two cases (`mapi` and `mapi_`) should produce consistent results. They should both produce errors. The problem is that your `Zipper` class is a generic class parameterized with the `TResult` type variable, but the `__call__` method within the class also uses another type variable called `TSource`. When you specialize a `Zipper` instance, only the `TResult` gets specialized. This leaves the `__call__` method not fully specialized. When you then call `compose`, there is an unspecified type variable that is trying to be matched. This should result in an error.

There was a bug in pyright that allowed the error to go unreported in the case of `mapi`. This will be fixed in the next release.

Note that mypy also generates errors when run on your code sample, but the error messages don't make sense to me. I suspect you're hitting a bug in mypy.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.97, which I just published. It will also be included in the next release of pylance.
